bin/runtime/src/migrations/custom_scheduler_migration.rs

```diff
@@ -0,0 +1,77 @@
+use codec::{Decode, Encode};
+use frame_support::{
+    log,
+    pallet_prelude::{Get, StorageVersion},
+    traits::OnRuntimeUpgrade,
+};
+use pallet_scheduler::{Agenda, Config, Pallet};
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+use crate::Weight;
+
+const TARGET: &str = "runtime::scheduler::migration";
+
+/// Custom migrations the scheduler pallet from V0 to V3 that only bumps StorageVersion to 3
+pub struct MigrateToV3<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let version = StorageVersion::get::<Pallet<T>>();
+        if version != 0 {
+            log::warn!(
+                target: TARGET,
+                "skipping v0 to v3 migration: executed on wrong storage version.\
+				Expected version 0, found {:?}",
+                version,
+            );
+            return T::DbWeight::get().reads(1);
+        }
+
+        let agendas = Agenda::<T>::iter_keys().count() as u32;
```

Comment:
nit: no idea whether this is somehow important. Line `38` suggests this call takes one read, but skipping through code suggests it can take up to `agendas` storage reads.

Reply:
Good catch! Since `Agenda` is `StorageMap`, `iter_keys` returns an iterator that is consumed by `count()`, which calls `next()` repeatedly until `None`. I adapted code so now correct weight is returned.

Reply:
It depends on how is the `Iterator` for `StorageMap` implemented. I just looked at `some` version and it looks like it `reads` one by one (no batching). Probably number of reads for that method is something like `agendas + 1` (+1 for reading `None` for next), so line should have +1 (but I don't know :)).